pkg/workflows/shardorchestrator/plugin.go

```diff
@@ -0,0 +1,235 @@
+	totalObservations := len(aos)
+	allWorkflows := []string{}
+
+	// Collect shard health observations and workflows
+	for _, ao := range aos {
+		observation := &pb.Observation{}
+		if err := proto.Unmarshal(ao.Observation, observation); err != nil {
+			p.lggr.Warnf("failed to unmarshal observation: %v", err)
+			continue
+		}
+
+		for shardID, healthy := range observation.Status {
+			if healthy {
+				currentShardHealth[shardID]++
+			}
+		}
+
+		// Collect workflow IDs
+		allWorkflows = append(allWorkflows, observation.Hashes...)
+	}
+
+	// Deduplicate workflows
+	workflowMap := make(map[string]bool)
+	for _, wf := range allWorkflows {
+		workflowMap[wf] = true
+	}
+	allWorkflows = make([]string, 0, len(workflowMap))
+	for wf := range workflowMap {
+		allWorkflows = append(allWorkflows, wf)
+	}
+	slices.Sort(allWorkflows) // Ensure deterministic order
+
+	// Determine desired shard count based on observations
+	healthyShardCount := uint32(0)
+	for shardID, count := range currentShardHealth {
+		if count > int(p.config.F) {
+			healthyShardCount++
+			// Update store with healthy shard
+			p.store.SetShardHealth(shardID, true)
+		}
+	}
+
+	// Ensure within bounds
+	if healthyShardCount < p.minShardCount {
+		healthyShardCount = p.minShardCount
+	}
+	if healthyShardCount > p.maxShardCount {
+		healthyShardCount = p.maxShardCount
+	}
+
+	// Use deterministic hashing to assign workflows to shards
+	routes := make(map[string]*pb.WorkflowRoute)
+	for _, wfID := range allWorkflows {
+		assignedShard := p.store.GetShardForWorkflow(wfID)
+		routes[wfID] = &pb.WorkflowRoute{
+			Shard: assignedShard,
+		}
+	}
+
+	// Update routing state
+	outcome := &pb.Outcome{
+		State: &pb.RoutingState{
+			Id: prevOutcome.State.Id + 1,
+			State: &pb.RoutingState_RoutableShards{
+				RoutableShards: healthyShardCount,
+			},
+		},
+		Routes: routes,
+	}
+
+	p.lggr.Infow("Consensus Outcome", "healthyShards", healthyShardCount, "totalObservations", totalObservations, "workflowCount", len(routes))
+
+	return proto.MarshalOptions{Deterministic: true}.Marshal(outcome)
+}
+
+func (p *Plugin) Reports(_ context.Context, _ uint64, outcome ocr3types.Outcome) ([]ocr3types.ReportPlus[[]byte], error) {
+	allOraclesTransmitNow := &ocr3types.TransmissionSchedule{
+		Transmitters:       make([]commontypes.OracleID, p.config.N),
+		TransmissionDelays: make([]time.Duration, p.config.N),
+	}
+
+	for i := 0; i < p.config.N; i++ {
+		allOraclesTransmitNow.Transmitters[i] = commontypes.OracleID(i)
+	}
+
+	info, err := structpb.NewStruct(map[string]any{
+		"keyBundleName": "evm",
+	})
+	if err != nil {
+		return nil, err
+	}
+	infoBytes, err := proto.MarshalOptions{Deterministic: true}.Marshal(info)
+	if err != nil {
+		return nil, err
+	}
+
+	return []ocr3types.ReportPlus[[]byte]{
+		{
+			ReportWithInfo: ocr3types.ReportWithInfo[[]byte]{
+				Report: types.Report(outcome),
+				Info:   infoBytes,
+			},
+			TransmissionScheduleOverride: allOraclesTransmitNow,
```

Comment:
I don't think we need this - we can use the transmission schedule from OCR config (no overrides).